rslib/src/decks/add.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+use super::name::{immediate_parent_name, normalize_native_name};
+use crate::{error::FilteredDeckError, prelude::*};
+use std::borrow::Cow;
+
+impl Collection {
+    /// Normalize deck name and rename if not unique. Bumps mtime and usn if
+    /// name was changed, but otherwise leaves it the same.
+    pub(super) fn prepare_deck_for_update(&mut self, deck: &mut Deck, usn: Usn) -> Result<()> {
+        if let Cow::Owned(name) = normalize_native_name(&deck.name) {
+            deck.name = name;
+            deck.set_modified(usn);
+        }
+        self.ensure_deck_name_unique(deck, usn)
+    }
+
+    /// Add or update an existing deck modified by the user. May add parents,
+    /// or rename children as required. Prefer add_deck() or update_deck() to
+    /// be explicit about your intentions; this function mainly exists so we
+    /// can integrate with older Python code that behaved this way.
+    pub(crate) fn add_or_update_deck(&mut self, deck: &mut Deck) -> Result<OpOutput<()>> {
+        if deck.id.0 == 0 {
+            self.add_deck(deck)
+        } else {
+            self.update_deck(deck)
+        }
+    }
+
+    /// Add a new deck. The id must be 0, as it will be automatically assigned.
+    pub fn add_deck(&mut self, deck: &mut Deck) -> Result<OpOutput<()>> {
+        if deck.id.0 != 0 {
+            return Err(AnkiError::invalid_input("deck to add must have id 0"));
+        }
+
+        self.transact(Op::AddDeck, |col| col.add_deck_inner(deck, col.usn()?))
+    }
+
+    pub(crate) fn add_deck_inner(&mut self, deck: &mut Deck, usn: Usn) -> Result<()> {
+        self.prepare_deck_for_update(deck, usn)?;
+        deck.set_modified(usn);
+        self.match_or_create_parents(deck, usn)?;
+        self.add_deck_undoable(deck)
+    }
+
+    pub fn update_deck(&mut self, deck: &mut Deck) -> Result<OpOutput<()>> {
+        self.transact(Op::UpdateDeck, |col| {
```

Comment:
Recently I've been trying to have an impl Collection {} at the top of the file with any public methods, and then another one for any crate/private methods, just to keep things a bit neater (eg tags/findreplace.rs)

Reply:
I'll do that then. 🙂